src/Reserve.sol

```diff
@@ -37,44 +122,146 @@ contract Reserve is IReserve, Ownable {
         _;
     }
 
-    function withdraw(
-        IERC20 token,
-        address to,
-        uint256 amt
-    ) public override onlyUser {
-        uint256 balance = withdrawable[token];
-        require(balance >= amt, "Reserve: withdrawal over balance");
-        withdrawable[token] = balance - amt;
-        require(token.transfer(to, amt), "Reserve: transfer failed");
-        emit Withdrawn(msg.sender, token, to, amt);
+    /// @notice Sets a plugin for a given token.
+    /// All future deposits and withdrawals of that token will be made using that plugin.
+    /// All currently deposited tokens of that type will be withdrawn from the plugin previously
+    /// set for that token and deposited into the new one.
+    /// If no plugin has been set, funds are deposited from the reserve itself.
+    /// If no plugin is being set, funds are deposited into the reserve itself.
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param newPlugin The new plugin address. `NO_PLUGIN` if no plugin should be used.
+    function setPlugin(IERC20 token, IReservePlugin newPlugin) public onlyOwner {
+        IReservePlugin oldPlugin = plugins[token];
+        plugins[token] = newPlugin;
+        uint256 amt = deposited[token];
+        _beforeWithdrawal(token, oldPlugin, amt);
+        _transfer(token, _pluginAddr(oldPlugin), _pluginAddr(newPlugin), amt);
+        _afterDeposition(token, newPlugin, amt);
+        emit PluginSet(msg.sender, token, oldPlugin, newPlugin);
     }
 
+    /// @notice Deposits funds into the reserve.
+    /// The reserve will `transferFrom` `amt` tokens from the `from` address.
+    /// Callable only by a current user.
+    /// @param token The used token.
+    /// @param from The address from which funds are deposited.
+    /// @param amt The deposited amount.
     function deposit(
         IERC20 token,
         address from,
         uint256 amt
     ) public override onlyUser {
-        withdrawable[token] += amt;
-        require(token.transferFrom(from, address(this), amt), "Reserve: transfer failed");
+        IReservePlugin plugin = plugins[token];
+        require(from != address(plugin) && from != address(this), "Reserve: deposition from self");
```

Comment:
`from` reserve would be a wild attack, good check. :)

Reply:
I'd be surprised to see `DripsHub` attacking the reserve, but it's still better to not have a built-in "feature" of doubling the balance and breaking the API :laughing: